pkg/sparrow/controller.go

```diff
@@ -0,0 +1,270 @@
+	}
+}
+
+// HandleErrors listens for errors and unregisters failed checks.
+func (cc *ChecksController) HandleErrors(ctx context.Context) {
+	log := logger.FromContext(ctx)
+
+	for {
+		select {
+		case err := <-cc.cErr:
+			var runErr *ErrRunningCheck
+			if errors.As(err, &runErr) {
+				uErr := cc.UnregisterCheck(ctx, runErr.Check)
+				if uErr != nil {
+					log.ErrorContext(ctx, "Failed to unregister check", "check", runErr.Check.Name(), "error", uErr)
+				}
+			}
+			log.ErrorContext(ctx, "Error while running check", "error", err)
+		case <-ctx.Done():
+			return
+		case <-cc.done:
+			cc.cErr <- nil
+			return
+		}
+	}
+}
+
+// Shutdown shuts down the ChecksController.
+func (cc *ChecksController) Shutdown(ctx context.Context) (err error) {
+	log := logger.FromContext(ctx)
+
+	for _, c := range cc.checks.Iter() {
+		cErr := cc.UnregisterCheck(ctx, c)
+		if cErr != nil {
+			log.Error("Failed to unregister check while shutting down", "error", cErr)
+			err = errors.Join(err, cErr)
+		}
+	}
+	cc.done <- struct{}{}
+	close(cc.done)
+	return err
+}
+
+// Reconcile reconciles the checks.
+// It registers new checks, updates existing checks and unregisters checks not in the new config.
+func (cc *ChecksController) Reconcile(ctx context.Context, cfg runtime.Config) {
+	log := logger.FromContext(ctx)
+
+	newChecks, err := factory.NewChecksFromConfig(cfg)
+	if err != nil {
+		log.ErrorContext(ctx, "Failed to create checks from config", "error", err)
+		return
+	}
+
+	// Update existing checks and create a list of checks to unregister
+	var unregList []checks.Check
+	for _, c := range cc.checks.Iter() {
+		conf := cfg.For(c.Name())
+		if conf != nil {
+			err = c.SetConfig(conf)
+			if err != nil {
+				log.ErrorContext(ctx, "Failed to set config for check", "check", c.Name(), "error", err)
+			}
+			delete(newChecks, c.Name())
+		} else {
+			unregList = append(unregList, c)
+		}
+	}
+
+	// Unregister checks not in the new config
+	for _, c := range unregList {
+		err = cc.UnregisterCheck(ctx, c)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to unregister check", "check", c.Name(), "error", err)
+		}
+	}
+
+	// Register new checks
+	for _, c := range newChecks {
+		err = cc.RegisterCheck(ctx, c)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to register check", "check", c.Name(), "error", err)
+		}
+	}
+}
+
+// RegisterCheck registers a new check.
+func (cc *ChecksController) RegisterCheck(ctx context.Context, check checks.Check) error {
+	log := logger.FromContext(ctx).With("check", check.Name())
+
+	// Add prometheus collectors of check to registry
+	for _, collector := range check.GetMetricCollectors() {
+		if err := cc.metrics.GetRegistry().Register(collector); err != nil {
+			log.ErrorContext(ctx, "Could not add metrics collector to registry")
+		}
+	}
+
+	go func() {
+		err := cc.runCheck(ctx, check)
+		if err != nil {
+			log.ErrorContext(ctx, "Failed to start check", "error", err)
+			cc.cErr <- err
+		}
+	}()
+
```

Comment:
This function, which is called RegisterCheck, does multiple things:

- It registers the check in the metrics registry (that's fine)
- It adds the check to the state (that's exactly what the name says)
- It starts running the check.

The latter is deeply problematic. To run the check, we start a goroutine in this function (the function will return in a bit but the goroutine will keep on going on). The goroutine calls another function (`runCheck`), which starts a separate goroutine, which finally (!!!) initiates the `Run` method of the `Check`. It then has a `for` to catch the results coming in. 

Still, we're not done. The `Run` method kinda does the same. It just has a `for` to catch the results of the internal implementation of how the check must run, which (suprise!) is a function which spawns multiple goroutines (which is fine IMO).

We have to watch out for multiple indirections. Too much abstraction will break our code. We should rethink how the checks start running, as this seems too convoluted, in my humble opinion. 

I think this can be addressed together with refactoring the `CheckController` to start with `Run` method and just do its thing (sorry, I sound like a broken record ATM :D ).

Reply:
I was like: "the check only runs once when it's registered and then it's over?". And then I was like "Oh no, the checks only start running when a new config comes in". And then I was like "The checks start running forever when registred but when a new check comes in, they are also updated".

I think we can find a better way to handle this :)

Reply:
I could also partly refactor it now to call `cc.runCheck` and then there the two go routines are triggered:
```go
func (cc *ChecksController) runCheck(ctx context.Context, check checks.Check) {
	go func() {
		err := check.Run(ctx)
		if err != nil {
			log.ErrorContext(ctx, "Failed to run check", "error", err)
			cc.cErr <- &ErrRunningCheck{
				Check: check,
				Err:   err,
			}
		}
	}()

	go func() {
		err := cc.fanInResults(ctx, check)
		if err != nil {
			log.ErrorContext(ctx, "Failed to fan in results", "error", err)
			cc.cErr <- err
		}
	}()
}
```
That way we can avoid creating a go routine in a go routine (at least in the controller).

Reply:
@puffitos and I had a call regarding this. We came to the following conclusion:

We'll remove the `ResultChan` function from the Checks interface again and reimplement a `fanInChecks` functionality so the controller initializes one Result Channel on its initialization and sets exactly this result channel for each check.

Reply:
I've refactored it in another branch and opened a PR to this branch to address the discussion: #106